lib/default.ml

```diff
@@ -38,27 +44,73 @@ module Str = struct
       Raise.Default.errorf
         ~loc:ptype_loc
         "can't derive default for an abstract type without a manifest"
-    | Some typ -> expr_from_core_type ~loc typ
+    | Some typ -> expr_from_core_type_exn ~loc typ
 
   let field_binding ~loc {pld_name; pld_type; _} =
+    let open Util.Result_ in
     let lident = {txt = Lident pld_name.txt; loc} in
-    let expr = expr_from_core_type ~loc pld_type in
+    expr_from_core_type ~loc pld_type >|= fun expr ->
     (lident, expr)
 
   let value_expr_from_labels ~loc labels =
+    let open Util.Result_ in
     let field_bindings = List.map (field_binding ~loc) labels in
+    Util.List_.all_ok field_bindings >|= fun field_bindings ->
     Ast_builder.Default.pexp_record ~loc field_bindings None
 
+  let value_expr_from_labels_exn ~loc labels =
```

Comment:
Why has this change led to all these new functions - they don't seem to be for variant types?

Reply:
The `value_expr_from_labels` is shared between the functions that return a default expression for a record and for a variant who's argument is a record.

When deriving a default value for a variant type we need to go through each constructor see if we can derive a generic default value for its arguments. Using a result makes this easier and more elegant. In the record case, if we can't derive a default for the record it's game over straigt away so we're using a wrapper that raise the error that will be displayed to the end user.

Does it make more sense now?

Reply:
Yep